dynamic_programming/regex_match.py

```diff
@@ -0,0 +1,106 @@
+"""
+Regex matching check if a text matches pattern or not.
+Pattern:
+    '.' Matches any single character.
+    '*' Matches zero or more of the preceding element.
+More info:
+    https://medium.com/trick-the-interviwer/regular-expression-matching-9972eb74c03
+"""
+
+
+def recursive_match(text: str, pattern: str) -> bool:
+    """
+    Recursive matching algorithm.
+
+    Time complexity: O(2 ^ (|text| + |pattern|))
+    Space complexity: Recursion depth is O(|text| + |pattern|).
+
+    :param text: Text to match.
+    :param pattern: Pattern to match.
+    :return: True if text matches pattern, False otherwise.
+
+    >>> recursive_match('abc', 'a.c')
+    True
+    >>> recursive_match('abc', 'af*.c')
+    True
+    >>> recursive_match('abc', 'a.c*')
+    True
+    >>> recursive_match('abc', 'a.c*d')
+    False
+    >>> recursive_match('aa', '.*')
+    True
+    """
+    if not text and not pattern:
+        return True
+
+    if text and not pattern:
+        return False
+
+    if not text and pattern and pattern[-1] != "*":
+        return False
+
+    if not text and pattern and pattern[-1] == "*":
+        return recursive_match(text, pattern[:-2])
```

Comment:
```suggestion
    if not text:
        return pattern[-1] == "*" and recursive_match(text, pattern[:-2])
```

1) No need to check if `pattern` is non-empty because the previous if-statement already handles it

2) If `pattern[-1] != "*"`, then the above code short-circuits and immediately returns False. Otherwise, it'll proceed to call `recursive_match` as usual.